pkg/util/vfsutil/copy.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package vfsutil
+
+import (
+	"io"
+
+	"github.com/cockroachdb/pebble/vfs"
+)
+
+// CopyRecursive recursively copies all files and directories from srcPath on
+// srcFS to dstPath on dstFS. If the destination path does not exist, it is
+// created.
+// Files that already exist in the destination path will be replaced.
+func CopyRecursive(srcFS, dstFS vfs.FS, srcPath, dstPath string) error {
+	srcInfo, err := srcFS.Stat(srcPath)
+	if err != nil {
+		return err
+	}
+
+	if srcInfo.IsDir() {
+		if err := dstFS.MkdirAll(dstPath, srcInfo.Mode()); err != nil {
+			return err
+		}
+
+		entries, err := srcFS.List(srcPath)
+		if err != nil {
+			return err
+		}
+
+		for _, entry := range entries {
+			srcEntryPath := srcFS.PathJoin(srcPath, entry)
+			dstEntryPath := dstFS.PathJoin(dstPath, entry)
+			if err := CopyRecursive(srcFS, dstFS, srcEntryPath, dstEntryPath); err != nil {
+				return err
+			}
```

Comment:
Should we fsync() the dstPath here to ensure newly created directory entries are also present?

Reply:
A follow-up here, if you do decide the fsync the directory, you might need some error handling since some filesystems return an error when fsyncing a directory.

Reply:
Oof, reasonable point, but I think getting into that would be a distraction at this point. I'll go ahead and merge this.